Contribution guidelines for `bevy_audio`

SolarLiner opened the pull request:
# Objective

Provide guidelines for contributing code to `bevy_audio`, with a focus on the critical sections of the audio engine.

## Changelog

Added to the crate-level documentation comment with a section introducing audio programming, real-time safety and why it is important to audio programming, as well as recommendations for some programming use-cases. The section concludes with links to more resources about audio programming.

I might have gone overboard with the writeup, but I didn't want to assume a lot out of potential `bevy_audio` contributors, and so I spent a bit of time defining terms as simply as I could. 

I didn't want to pressure people to do so, but the first link on the additional resources should really be "required reading" as it goes more in depth about the why and how of audio programming.

alice-i-cecile reviewed (approved):
Really nice stuff. I think this is about the right level of detail and assumes a competent audience that's new to audio engineering :)

I think it might make more sense to rebrand this as something like "fundamentals of working with audio" or similar: it's useful to end users diving into low-level audio code of their own.

SolarLiner commented:
> I think it might make more sense to rebrand this as something like "fundamentals of working with audio" or similar: it's useful to end users diving into low-level audio code of their own.

Do you mean that we should move this out of the crate docs then? Or that I rename the top-level heading to "Fundamentals of working with audio"?

rparrett commented:
I am not an audio person and can't really review the text.

But as a casual contributor to `bevy_audio`, I am having some trouble relating to this. Maybe Alice's rebranding will help reframe things, but I feel like this is missing a lot of "bevy context." I have the feeling that the vast majority of the text would apply more to contributors of `rodio` or `cpal`. I don't immediately know what sort of audio thread stuff Bevy is doing. This is describing a lot of specific data structures / threading concurrency strategies and I have no idea if this talk is sort of theoretical about what we should be doing, or something that is already happening in rodio, or in cpal, or what. I don't think it's happening in Bevy.

Even in the section that refers specifically to Bevy, we have

```
//! For the Bevy engine, that is:
//!
//! - In the callback of `cpal::Stream::build_input_stream` and `cpal::Stream::build_output_stream`, and all
//!   functions called from them
//! - In implementations of the [`Source`] trait, and all functions called from it
```

But I don't think there's an implementation of `Source` in Bevy, other than in an example. Likewise, `build_input_stream` and `build_output_stream` don't seem to appear in the engine.

james7132 commented on crates/bevy_audio/src/lib.rs:
````diff
@@ -19,6 +19,108 @@
 //! }
 //! ```
+//!
+//! # Contributing
````
This is user-facing documentation. This is likely more suited for a crate level CONTRIBUTING.md, or as a non-doc comment. The overall design philosophy is probably worth mentioning in passing or tucking away in a public but lower level module, but much of this are implementation details the average user shouldn't need to think about when using the engine.

SolarLiner replied:
Just saw your comment after pushing some changes. I do agree with this, putting it in the crate-level doc comments was something @alice-i-cecile suggested, but any place for it is fine by me. `CONTRIBUTING.md` is a good alternative place I think.

james7132 reviewed (changes requested):
Overall, this is great documentation and general advice for developing audio within the engine, but documentation must take into account the target demographic of who is consuming it.

SolarLiner commented:
Changed the wording of the main heading, and moved the Bevy-specific paragraph up, as well as adding a "foreword" about the documentation that is following.

SolarLiner commented:
Moved the documentation text to a separate CONTRIBUTING.md file. It's less visible than the crate documentation, but also only apply to contributions to the code, instead of user-facing documentation. I thought of mentioning adding effects and mentioning the real-time audio section, but since currently with `rodio` it's not easy to write audio effects, or even apply them to sources or globally, I ultimately decided against it.

This is something we can always add on a second pass, after having fleshed out the effect processing story (i.e. probably after having moved away from `rodio`).
